cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java

```diff
@@ -34,24 +36,38 @@
  *   <li>For sorting of language literals we first sort on language, next on value</li>
  *   <li>For sorting of typed literals we first sort on datatype, next on value</li>
  * </ul>
- *
  */
 public class StatementSorter implements Function<Model, List<Statement>> {
 
   @Override
   public List<Statement> apply(Model model) {
     List<Statement> statements = getCanonicalStatements(model);
     statements.sort(getStatementComparator());
+    statements = removeDuplicateStatements(statements);
     return statements;
   }
 
   private List<Statement> getCanonicalStatements(Model model) {
     List<Statement> statements = new ArrayList<>();
-    SortedBlock root = new SortedBlock(model);
+    SortedBlock root = SortedBlock.create(model);
     processBlock(statements, null, root);
     return statements;
   }
 
+  private List<Statement> removeDuplicateStatements(List<Statement> statements) {
+    for (int i = 0; i < statements.size() - 1; i++) {
+      Statement statement = statements.get(i);
+      Statement nextStatement = statements.get(i + 1);
+      boolean isSame = getStatementComparator().compare(statement, nextStatement) == 0;
+      if (isSame) {
+        statements.remove(i + 1);
+        i -= 1;
+      }
+    }
+
+    return statements;
+  }
+
```

Comment:
Detection and removal of duplicates.

Reply:
Wouldn't casting to set work as well? 🤔

Reply:
true. I was not sure either but decided to return it so we can change the implementation later. @michielrogissart-cognizone